sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -0,0 +1,3597 @@
+  // here. Setting it to 256 gave best possible performance for several
+  // benchmarks.
+  static const pi_uint32 MaxNumEventsPerPool = [] {
+    const auto MaxNumEventsPerPoolEnv =
+        std::getenv("ZE_MAX_NUMBER_OF_EVENTS_PER_EVENT_POOL");
+    return MaxNumEventsPerPoolEnv ? std::atoi(MaxNumEventsPerPoolEnv) : 256;
+  }();
+
+  if (MaxNumEventsPerPool == 0) {
+    zePrint("Zero size can't be specified in the "
+            "ZE_MAX_NUMBER_OF_EVENTS_PER_EVENT_POOL\n");
+    return ZE_RESULT_ERROR_INVALID_SIZE;
+  }
+
+  Index = 0;
+  // Create one event ZePool per MaxNumEventsPerPool events
+  if ((ZeEventPool == nullptr) ||
+      (NumEventsAvailableInEventPool[ZeEventPool] == 0)) {
+    // Creation of the new ZePool with record in NumEventsAvailableInEventPool
+    // and initialization of the record in NumEventsLiveInEventPool must be done
+    // atomically. Otherwise it is possible that decrementAliveEventsInPool will
+    // be called for the record in NumEventsLiveInEventPool before its
+    std::lock(NumEventsAvailableInEventPoolMutex,
+              NumEventsLiveInEventPoolMutex);
+    std::lock_guard<std::mutex> NumEventsAvailableInEventPoolGuard(
+        NumEventsAvailableInEventPoolMutex, std::adopt_lock);
+    std::lock_guard<std::mutex> NumEventsLiveInEventPoolGuard(
+        NumEventsLiveInEventPoolMutex, std::adopt_lock);
+
+    ze_event_pool_desc_t ZeEventPoolDesc;
+    ZeEventPoolDesc.count = MaxNumEventsPerPool;
+    ZeEventPoolDesc.flags = ZE_EVENT_POOL_FLAG_TIMESTAMP;
+    ZeEventPoolDesc.version = ZE_EVENT_POOL_DESC_VERSION_CURRENT;
+
+    ze_device_handle_t ZeDevice = Device->ZeDevice;
+    if (ze_result_t ZeRes =
+            zeEventPoolCreate(Device->Platform->ZeDriver, &ZeEventPoolDesc, 1,
+                              &ZeDevice, &ZeEventPool))
+      return ZeRes;
+    NumEventsAvailableInEventPool[ZeEventPool] = MaxNumEventsPerPool - 1;
+    NumEventsLiveInEventPool[ZeEventPool] = MaxNumEventsPerPool;
+  } else {
+    std::lock_guard<std::mutex> NumEventsAvailableInEventPoolGuard(
+        NumEventsAvailableInEventPoolMutex);
+    Index = MaxNumEventsPerPool - NumEventsAvailableInEventPool[ZeEventPool];
+    --NumEventsAvailableInEventPool[ZeEventPool];
+  }
+  ZePool = ZeEventPool;
+  return ZE_RESULT_SUCCESS;
+}
+
+ze_result_t
+_pi_context::decrementAliveEventsInPool(ze_event_pool_handle_t ZePool) {
+  std::lock_guard<std::mutex> Lock(NumEventsLiveInEventPoolMutex);
+  --NumEventsLiveInEventPool[ZePool];
+  if (NumEventsLiveInEventPool[ZePool] == 0) {
+    return zeEventPoolDestroy(ZePool);
+  }
+  return ZE_RESULT_SUCCESS;
+}
+
+// Some opencl extensions we know are supported by all Level0 devices.
+constexpr char ZE_SUPPORTED_EXTENSIONS[] =
+    "cl_khr_il_program cl_khr_subgroups cl_intel_subgroups "
+    "cl_intel_subgroups_short cl_intel_required_subgroup_size ";
+
+// Map L0 runtime error code to PI error code
+static pi_result mapError(ze_result_t ZeResult) {
```

Comment:
I think it's better to use `map` or `unordered_map`. If the `ZE_*` values are sequential and pretty low you could employ an array.

Reply:
Done.